From 4d48959f103cf997ee688e0da846a1c192785a16 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:36:48 -0400
Subject: [virt] kvm: remove .done from struct kvm_async_pf

Message-id: <1406040016-3289-14-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86092
O-Subject: [RHEL7 PATCH v2 013/221] kvm: remove .done from struct kvm_async_pf
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Radim Krcmar <rkrcmar@redhat.com>

'.done' is used to mark the completion of 'async_pf_execute()', but
'cancel_work_sync()' returns true when the work was canceled, so we
use it instead.

Signed-off-by: Radim Krcmar <rkrcmar@redhat.com>
Reviewed-by: Paolo Bonzini <pbonzini@redhat.com>
Reviewed-by: Gleb Natapov <gleb@redhat.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 98fda169290b3b28c0f2db2b8f02290c13da50ef)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/kvm_host.h b/include/linux/kvm_host.h
index f029f55..0fe5a79 100644
--- a/include/linux/kvm_host.h
+++ b/include/linux/kvm_host.h
@@ -190,7 +190,6 @@ struct kvm_async_pf {
  unsigned long addr;
  struct kvm_arch_async_pf arch;
  struct page *page;
- bool done;
 };
 
 void kvm_clear_async_pf_completion_queue(struct kvm_vcpu *vcpu);
diff --git a/virt/kvm/async_pf.c b/virt/kvm/async_pf.c
index 8a39dda..b197950 100644
--- a/virt/kvm/async_pf.c
+++ b/virt/kvm/async_pf.c
@@ -75,7 +75,6 @@ static void async_pf_execute(struct work_struct *work)
  spin_lock(&vcpu->async_pf.lock);
  list_add_tail(&apf->link, &vcpu->async_pf.done);
  apf->page = page;
- apf->done = true;
  spin_unlock(&vcpu->async_pf.lock);
 
  /*
@@ -99,9 +98,8 @@ void kvm_clear_async_pf_completion_queue(struct kvm_vcpu *vcpu)
   struct kvm_async_pf *work =
    list_entry(vcpu->async_pf.queue.next,
        typeof(*work), queue);
-  cancel_work_sync(&work->work);
   list_del(&work->queue);
-  if (!work->done) { /* work was canceled */
+  if (cancel_work_sync(&work->work)) {
    mmdrop(work->mm);
    kvm_put_kvm(vcpu->kvm); /* == work->vcpu->kvm */
    kmem_cache_free(async_pf_cache, work);
@@ -166,7 +164,6 @@ int kvm_setup_async_pf(struct kvm_vcpu *vcpu, gva_t gva, gfn_t gfn,
   return 0;
 
  work->page = NULL;
- work->done = false;
  work->vcpu = vcpu;
  work->gva = gva;
  work->addr = gfn_to_hva(vcpu->kvm, gfn);
-- 
1.7.1